### PR TITLE
Savotina Valeria, Lab №4, var: 2

### PR DESCRIPTION
**Pull request by Savotina**
This pass can merge multiplication and addition into a single match.fma

### CODE REVIEW

File: mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    auto replaceAndEraseOp = [&](mlir::LLVM::FMulOp &mulOp,
+                                 mlir::LLVM::FAddOp &addOp,
+                                 mlir::Value &thirdOperand) -> void {
+      builder.setInsertionPoint(addOp);
+      mlir::Value fmaResult =
+          builder.create<mlir::math::FmaOp>(addOp.getLoc(), mulOp.getOperand(0),
+                                            mulOp.getOperand(1), thirdOperand);
+      addOp.replaceAllUsesWith(fmaResult);
+      addOp.erase();
+      mulOp.erase();
```
Comment (aobolensk) [inline mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp:34]: You cannot remove mulOp every time. You need to check for uses of mulOp. You may see this example https://github.com/NN-complr-tech/llvm-nnsu-2024/pull/152#discussion_r1605668496

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp (hunk 1 above)
Comment (Savotina) [reply mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp:34]: @aobolensk, I don't remove mulOp every time, I check if the result of multiplication is used somewhere else before the addition operation with **isSingleUse**. I added tests relying on example, you can make sure

**Review by m-ly4 — APPROVED** [review]

**Comment by m-ly4** [comment]
This pass can merge multiplication and addition into a single match.fma

**Review by aobolensk — APPROVED** [review]

File: mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp (hunk 1 above)
Comment (aobolensk) [reply mlir/lib/Transforms/lab4/savorina_valeria/SavotinaMulAddPass.cpp:34]: Missed that point, thanks for clarification. Approved

**Pull request merged**